### PR TITLE
Healthchecks

### PR DESCRIPTION
**Pull request by mauri**
don't merge

### CODE REVIEW

File: api/src/main/thrift/org/apache/aurora/gen/api.thrift [hunk 1]
```diff
@@ -255,6 +255,54 @@ struct KillPolicy {
...
+// the command is logged to stdout and stderr. 
+struct ShellHealthChecker {  
+  1: string command
+}
+
+/** Describe a Mesos command Health check */
+struct HealthCheck {
+
+  // Amount of time to wait to start health checking the task after it
+  // transitions to `TASK_RUNNING` or `TASK_STATING` if the latter is
```
Comment (NDonatucci) [inline api/src/main/thrift/org/apache/aurora/gen/api.thrift]: Starting or staging?

File: src/main/java/org/apache/aurora/scheduler/configuration/ConfigurationManager.java [hunk 2]
```diff
@@ -402,6 +402,10 @@ public ITaskConfig validateAndPopulate(ITaskConfig config) throws TaskDescriptio
     if (config.isSetKillPolicy() && config.getKillPolicy().getGracePeriodSecs() < 0) {
       throw new TaskDescriptionException("Grace Period should be greater than 0");
     }
+    
+    if (config.isSetHealthCheck() && !config.getHealthCheck().isSetHttp() && !config.getHealthCheck().isSetShell()) {
```
Comment (NDonatucci) [inline src/main/java/org/apache/aurora/scheduler/configuration/ConfigurationManager.java]: What if both health checks are set?

File: src/main/java/org/apache/aurora/scheduler/mesos/DockerContainerTasks.java [hunk 3]
```diff
@@ -23,19 +24,19 @@
...
+	/** Configure the {@link Builder} based on the given {@link org.apache.aurora.scheduler.storage.entities.IAssignedTask */
 	public static void configureTask(
 			IAssignedTask task,
-			ITaskConfig taskConfig,
 			Builder taskBuilder,
-			AcceptedOffer acceptedOffer, IServerInfo serverInfo) {
-		LOG.info("Setting DOCKER Task. {}", taskConfig.getExecutorConfig().getData());
-		LOG.info("Instance. {}", taskConfig.getInstances());
+			IServerInfo serverInfo) {
+		LOG.info("Setting DOCKER Task. {}. Instances {}", task.getTask().getExecutorConfig().getData(),
```
Comment (NDonatucci) [inline src/main/java/org/apache/aurora/scheduler/mesos/DockerContainerTasks.java]: Maybe extract to a variable?

File: src/main/java/org/apache/aurora/scheduler/mesos/DockerContainerTasks.java [hunk 4]
```diff
@@ -73,12 +74,33 @@ public static void configureTask(
...
+			HealthCheck.Builder mesosHealthCheck = HealthCheck.newBuilder()
+					.setConsecutiveFailures(healthCheck.getConsecutiveFailures())
+					.setDelaySeconds(healthCheck.getDelaySeconds())
+					.setGracePeriodSeconds(healthCheck.getGracePeriodSeconds())
+					.setIntervalSeconds(healthCheck.getIntervalSeconds())
+					.setTimeoutSeconds(healthCheck.getTimeoutSeconds());
+			if (healthCheck.isSetShell()) {
+				mesosHealthCheck.setCommand(CommandInfo.newBuilder()
+						.setValue(healthCheck.getShell().getCommand()).build());
+			} else if (healthCheck.isSetHttp()) {
```
Comment (NDonatucci) [inline src/main/java/org/apache/aurora/scheduler/mesos/DockerContainerTasks.java:99]: If validation is done before then the "else" is not needed

File: src/main/python/apache/aurora/config/thrift.py [hunk 5]
```diff
@@ -355,3 +359,19 @@ def not_empty_or(item, default):
       cronCollisionPolicy=select_cron_policy(job.cron_collision_policy()),
       taskConfig=task,
       instanceCount=numberOfInstances)
+
+def convert_health_check(job_health_check):
+    print ">.HealthCheck: {}\n".format(job_health_check)
```
Comment (NDonatucci) [inline src/main/python/apache/aurora/config/thrift.py]: remove this :)

File: src/main/python/apache/aurora/config/thrift.py [hunk 6]
```diff
@@ -355,3 +359,19 @@ def not_empty_or(item, default):
...
+    hc.delaySeconds= fully_interpolated(job_health_check.initial_interval_secs())
+    hc.interval_seconds = fully_interpolated(job_health_check.interval_secs())
+    hc.max_consecutive_failures = fully_interpolated(job_health_check.max_consecutive_failures())
+    if job_health_check.health_checker().http() is not Empty:
+      hc.http = HttpHealthChecker(endpoint=fully_interpolated(job_health_check.health_checker().http().endpoint()))
+    elif job_health_check.health_checker().shell() is not Empty:
+      hc.shell = ShellHealthChecker(command=fully_interpolated(job_health_check.health_checker().shell().shell_command()))
+    else:
+      raise InvalidConfig("Config doesn't contain a valid HealthChecker. Include an http or shell checker.")
+    print "<.HealthCheck: {}\n".format(hc)
```
Comment (NDonatucci) [inline src/main/python/apache/aurora/config/thrift.py]: same here

File: src/main/python/apache/aurora/config/thrift.py [hunk 7]
```diff
@@ -285,6 +288,7 @@ def not_empty_or(item, default):
       fully_interpolated(task_raw.resources().disk())))
 
   task.killPolicy = KillPolicy(gracePeriodSecs=fully_interpolated(task_raw.finalization_wait()))
+  task.healthCheck = convert_health_check(job.health_check_config())
```
Comment (NDonatucci) [inline src/main/python/apache/aurora/config/thrift.py]: move this to job components part

File: src/test/java/org/apache/aurora/scheduler/TaskStatusHandlerImplTest.java [hunk 8]
```diff
@@ -233,7 +234,7 @@ public void testSuppressUnregisteredExecutorMessage() throws Exception {
         TASK_ID_A,
         Optional.absent(),
         KILLED,
-        Optional.absent()))
+        Optional.of("Service not Healthy")))
```
Comment (NDonatucci) [inline src/test/java/org/apache/aurora/scheduler/TaskStatusHandlerImplTest.java]: This means that executor_unregistered implies health checks failing in this example?

File: src/main/python/apache/aurora/config/thrift.py (hunk 7 above)
Comment (NDonatucci) [reply src/main/python/apache/aurora/config/thrift.py]: Also, how does this work when health_check is not defined?

File: src/main/python/apache/aurora/config/thrift.py [hunk 9]
```diff
@@ -355,3 +359,19 @@ def not_empty_or(item, default):
       cronCollisionPolicy=select_cron_policy(job.cron_collision_policy()),
       taskConfig=task,
       instanceCount=numberOfInstances)
+
+def convert_health_check(job_health_check):
```
Comment (NDonatucci) [inline src/main/python/apache/aurora/config/thrift.py:364]: Might want to raise an Invalid Config if both health_ckecks are set

File: src/main/java/org/apache/aurora/scheduler/configuration/ConfigurationManager.java (hunk 2 above)
Comment (juanfresia) [reply src/main/java/org/apache/aurora/scheduler/configuration/ConfigurationManager.java]: You need both of them to set up a Healthcheck, right? Should the error message read http _and_ shell?

File: api/src/main/thrift/org/apache/aurora/gen/api.thrift (hunk 1 above)
Comment (mauri) [reply api/src/main/thrift/org/apache/aurora/gen/api.thrift]: http://mesos.apache.org/api/latest/java/org/apache/mesos/Protos.HealthCheckOrBuilder.html#getDelaySeconds-- :)

File: src/test/java/org/apache/aurora/scheduler/TaskStatusHandlerImplTest.java (hunk 8 above)
Comment (mauri) [reply src/test/java/org/apache/aurora/scheduler/TaskStatusHandlerImplTest.java]: I thought about only setting that if the task has a healthcheck configured..

File: src/main/java/org/apache/aurora/scheduler/mesos/DockerContainerTasks.java (hunk 4 above)
Comment (mauri) [reply src/main/java/org/apache/aurora/scheduler/mesos/DockerContainerTasks.java:99]: but it makes the intention clear without needing to dig into the validator

File: src/test/java/org/apache/aurora/scheduler/storage/testing/StorageEntityUtil.java [hunk 10]
```diff
@@ -40,6 +40,7 @@ private StorageEntityUtil() {
   }
 
   private static void assertFullyPopulated(String name, Object object, Set<Field> ignoredFields) {
+    System.out.printf("%s - %s%n", name, object);
```
Comment (NDonatucci) [inline src/test/java/org/apache/aurora/scheduler/storage/testing/StorageEntityUtil.java]: remove

File: src/main/python/apache/aurora/config/schema/base.py [hunk 11]
```diff
@@ -181,7 +181,7 @@ class MesosJob(Struct):
   max_task_failures          = Default(Integer, 1)
   production                 = Default(Boolean, False)
   priority                   = Default(Integer, 0)
-  health_check_config        = Default(HealthCheckConfig, HealthCheckConfig())
+  health_check_config        = HealthCheckConfig #, HealthCheckConfig())
```
Comment (NDonatucci) [inline src/main/python/apache/aurora/config/schema/base.py]: remove comment

File: src/main/java/org/apache/aurora/scheduler/mesos/DockerContainerTasks.java [hunk 12]
```diff
@@ -23,16 +25,18 @@
 public class DockerContainerTasks {
     private static final Logger LOG = LoggerFactory.getLogger(DockerContainerTasks.class);
 
+	/** Configure the {@link Builder} based on the given {@link org.apache.aurora.scheduler.storage.entities.IAssignedTask */
```
Comment (juanfresia) [inline src/main/java/org/apache/aurora/scheduler/mesos/DockerContainerTasks.java]: Missing closing } ?

File: src/main/python/apache/aurora/config/schema/base.py (hunk 11 above)
Comment (juanfresia) [inline src/main/python/apache/aurora/config/schema/base.py]: does this change broke any client-side tests?

**Review by NDonatucci — APPROVED** [review]

**Pull request merged**